### PR TITLE
Add Schema Registry Avro functions for column deserialization

### PR DESCRIPTION
**Pull request by arerlend**
Thanks for contributing! We appreciate it :) 

For a Pull Request to be accepted, you must:
- Run scalafmt on your code using the `.scalafmt.conf` present in this project
- All tests must pass when you run `mvn clean test` 

Just in case, here are some tips that could prove useful when opening a pull request:
- Read the [Contributor's Guide](CONTRIBUTING.md)
- Make the title of pull request is clear and informative. 
- There should be a small number of commits, all with informative messages. 
- The pull request shouldn't introduce any breaking changes (unless will occur on the next release) 
- Any public code should be properly documented 
- Be sure to write tests for any changes in the pull request
- The code should build without any errors

### CODE REVIEW

File: schemaregistry-avro/README.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+However, there are two edge cases that will be common in streaming scenarios in which we are concerned with schema evolution -
+- Stream jobs reading old data with new schemas - only backwards compatible data will be readable, meaning that fields may be null.
+- Stream jobs reading new data with old schemas - even if the Spark job schema is forwards compatible with the new schema, projecting data written with the new schema to the old one will result in data loss in the case of additional fields being added.
+
+To handle the more dangerous second case, Spark functions will throw if incoming data contains fields that cannot be captured by the existing schema.  This behavior is based on the assumption that perceived data loss is prohibited.
+
+To handle the first first case, a parameter will be introduced called `requireExactSchemaMatch`:
+- If true, if the schema in the payload is not an exact match to the Spark-specified schema, then the job will throw.  This allows users to specify that their pipeline contain one schema only.  (q: add list of permitted guids?)
```
Comment (nyaghma) [inline schemaregistry-avro/README.md:75]: Remove the comment in ()

File: schemaregistry-avro/src/main/scala/com/microsoft/azure/schemaregistry/spark/avro/AvroDataToCatalyst.scala [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+     options: Map[java.lang.String, java.lang.String],
+     requireExactSchemaMatch: Boolean)
+  extends UnaryExpression with ExpectsInputTypes {
+
+  override def inputTypes: Seq[BinaryType] = Seq(BinaryType)
+
+  override lazy val dataType: DataType = {
+    val dt = SchemaConverters.toSqlType(new Schema.Parser().parse(expectedSchemaString)).dataType;
+    dt
+    // todo: compare stream compat to group compat and verify equal
```
Comment (nyaghma) [inline schemaregistry-avro/src/main/scala/com/microsoft/azure/schemaregistry/spark/avro/AvroDataToCatalyst.scala:48]: There is a "todo" comment here. Is it a future work or should it be done now?

File: schemaregistry-avro/src/main/scala/com/microsoft/azure/schemaregistry/spark/avro/functions.scala [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+       data: Column,
+       schemaId: String,
+       clientOptions: java.util.Map[java.lang.String, java.lang.String],
+       requireExactSchemaMatch: Boolean = true): Column = {
+    new Column(AvroDataToCatalyst(data.expr, schemaId, clientOptions.asScala.toMap, requireExactSchemaMatch))
+  }
+
+  /***
+   * Incoming
+   * @param data
```
Comment (nyaghma) [inline schemaregistry-avro/src/main/scala/com/microsoft/azure/schemaregistry/spark/avro/functions.scala:54]: Update the param & return descriptions

File: schemaregistry-avro/src/main/scala/com/microsoft/azure/schemaregistry/spark/avro/functions.scala [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+   * @param props
+   * @return
+   */
+  def to_avro(data: Column, props : Map[String, AnyRef]): Column = {
```
Comment (nyaghma) [inline schemaregistry-avro/src/main/scala/com/microsoft/azure/schemaregistry/spark/avro/functions.scala:58]: props parameter is not being used. Why is this needed if it just returns the input data without any modifications?

File: schemaregistry-avro/src/main/scala/com/microsoft/azure/schemaregistry/spark/avro/AvroDataToCatalyst.scala (hunk 2 above)
Comment (arerlend) [reply schemaregistry-avro/src/main/scala/com/microsoft/azure/schemaregistry/spark/avro/AvroDataToCatalyst.scala:48]: Let's remove it.  The idea here is that stream should expect to see schemas that evolve the same way that the group evolves.  There are some nuances here though, specifically that a) stream will also throw to prevent data loss and b) stream can be configured to only match exactly to the incoming messages' schemas.  These conditions are slightly different from compatibility - stream compat setting was an older version of the idea.

File: schemaregistry-avro/src/main/scala/com/microsoft/azure/schemaregistry/spark/avro/functions.scala (hunk 3 above)
Comment (arerlend) [reply schemaregistry-avro/src/main/scala/com/microsoft/azure/schemaregistry/spark/avro/functions.scala:54]: Changing to throw UnsupportedOperationException for now since it's not critical path.

File: schemaregistry-avro/src/main/scala/com/microsoft/azure/schemaregistry/spark/avro/functions.scala (hunk 4 above)
Comment (arerlend) [reply schemaregistry-avro/src/main/scala/com/microsoft/azure/schemaregistry/spark/avro/functions.scala:58]: Same thing as previous comment - changing to throw UnsupportedOperationException for now since it's not critical path.

**Comment by arerlend** [comment]
Permissive mode support is missing.

**Comment by tfayyaz** [comment]
Are there any updates on when this PR will be merged? I am coming across the same issue mentioned here where the stream does not initialise - https://github.com/arerlend/azure-schemaregistry-spark/issues/3

**Comment by kevinlacire** [comment]
Any updates ? this feature is a MVP in a pub/sub world
